File: sdk/core/src/main/java/com/securegateway/core/Crypto.java
package com.securegateway.core;

import com.goterl.lazysodium.LazySodium;
import com.goterl.lazysodium.interfaces.AEAD;
import com.goterl.lazysodium.interfaces.DiffieHellman;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.ServiceLoader;

/**
 * The E2EE crypto primitives (PRD FR-5), matching the Go reference
 * {@code internal/e2ee/e2ee.go} byte-for-byte. The interop vectors in
 * {@code internal/e2ee/testdata/vectors.json} are the cross-platform contract.
 *
 * <p>Scheme: X25519 ECDH (raw {@code crypto_scalarmult}, low-order points rejected)
 * &rarr; HKDF-SHA256 directional keys (see {@link Session}) &rarr; XChaCha20-Poly1305
 * IETF (24-byte nonce) with {@code wire = nonce(24) || ciphertext+tag(16)} and
 * {@code aad = utf8(id) || bigEndianUint64(ts)}.
 *
 * <p>X25519 and the AEAD come from native libsodium via lazysodium (the only
 * widely-available true 24-byte-nonce XChaCha20-Poly1305). HKDF is done with
 * {@link Hkdf} (RFC 5869 over HmacSHA256), since libsodium has no HKDF.
 */
public final class Crypto {

    /** X25519 key length and derived AEAD key length. */
    public static final int KEY_SIZE = 32;
    /** Per-session handshake nonce length each side contributes to the HKDF salt. */
    public static final int HANDSHAKE_NONCE_SIZE = 32;
    /** XChaCha20-Poly1305 nonce length, prepended to the ciphertext. */
    public static final int NONCE_SIZE = 24;
    /** Poly1305 authentication tag length. */
    public static final int TAG_SIZE = 16;

    static final String INFO_PREFIX = "secure-gateway/e2ee/v1|";
    static final String DIR_M2D = "m2d"; // mobile -> desktop
    static final String DIR_D2M = "d2m"; // desktop -> mobile

    // lazysodium is thread-safe (stateless native calls); share one instance. The
    // concrete binding (LazySodiumJava on the JVM, LazySodiumAndroid on device) is
    // supplied by the platform's registered SodiumProvider (see that interface).
    private static final LazySodium SODIUM = ServiceLoader.load(SodiumProvider.class)
            .findFirst()
            .orElseThrow(() -> new IllegalStateException(
                    "crypto: no SodiumProvider registered on the classpath "
                            + "(META-INF/services/com.securegateway.core.SodiumProvider)"))
            .get();
    private static final SecureRandom RANDOM = new SecureRandom();

    private Crypto() {
    }

    static LazySodium sodium() {
        return SODIUM;
    }

    /** Generate a fresh X25519 keypair. */
    public static KeyPair generateKeyPair() {
        byte[] priv = new byte[KEY_SIZE];
        RANDOM.nextBytes(priv);
        return new KeyPair(priv, publicFromPrivate(priv));
    }

    /** Derive the X25519 public key for {@code priv} (== Go {@code PublicFromPrivate}). */
    public static byte[] publicFromPrivate(byte[] priv) {
        requireLen(priv, KEY_SIZE, "private key");
        byte[] pub = new byte[KEY_SIZE];
        if (!((DiffieHellman.Native) SODIUM).cryptoScalarMultBase(pub, priv)) {
            throw new IllegalStateException("crypto: scalarmult_base failed");
        }
        return pub;
    }

    /** Fresh 32-byte session handshake nonce. */
    public static byte[] newHandshakeNonce() {
        byte[] n = new byte[HANDSHAKE_NONCE_SIZE];
        RANDOM.nextBytes(n);
        return n;
    }

    /**
     * X25519 ECDH shared secret. Mirrors Go's low-order-point rejection: an
     * all-zero result (low-order peer public key) is rejected.
     */
    public static byte[] deriveSharedSecret(byte[] myPriv, byte[] peerPub) {
        requireLen(myPriv, KEY_SIZE, "private key");
        requireLen(peerPub, KEY_SIZE, "peer public key");
        byte[] shared = new byte[KEY_SIZE];
        boolean ok = ((DiffieHellman.Native) SODIUM).cryptoScalarMult(shared, myPriv, peerPub);
        if (!ok || isAllZero(shared)) {
            throw new IllegalArgumentException("crypto: ecdh failed (low-order point)");
        }
        return shared;
    }

    /**
     * XChaCha20-Poly1305 seal with an explicit nonce. Returns the AEAD ciphertext
     * (with appended tag) only — callers prepend the nonce to form the wire bytes.
     */
    static byte[] aeadEncrypt(byte[] key, byte[] nonce, byte[] aad, byte[] plaintext) {
        requireLen(key, KEY_SIZE, "key");
        requireLen(nonce, NONCE_SIZE, "nonce");
        byte[] pt = plaintext == null ? new byte[0] : plaintext;
        byte[] ct = new byte[pt.length + TAG_SIZE];
        long[] ctLen = new long[1];
        boolean ok = ((AEAD.Native) SODIUM).cryptoAeadXChaCha20Poly1305IetfEncrypt(
                ct, ctLen, pt, pt.length, aad, aad.length, null, nonce, key);
        if (!ok) {
            throw new IllegalStateException("crypto: aead encrypt failed");
        }
        if (ctLen[0] != ct.length) {
            ct = Arrays.copyOf(ct, (int) ctLen[0]);
        }
        return ct;
    }

    /** XChaCha20-Poly1305 open. {@code ct} is ciphertext+tag (no nonce). */
    static byte[] aeadDecrypt(byte[] key, byte[] nonce, byte[] aad, byte[] ct) {
        requireLen(key, KEY_SIZE, "key");
        requireLen(nonce, NONCE_SIZE, "nonce");
        if (ct.length < TAG_SIZE) {
            throw new IllegalArgumentException("crypto: ciphertext too short");
        }
        byte[] pt = new byte[ct.length - TAG_SIZE];
        long[] ptLen = new long[1];
        boolean ok = ((AEAD.Native) SODIUM).cryptoAeadXChaCha20Poly1305IetfDecrypt(
                pt, ptLen, null, ct, ct.length, aad, aad.length, nonce, key);
        if (!ok) {
            throw new IllegalArgumentException("crypto: aead open failed (tampered or wrong key)");
        }
        if (ptLen[0] != pt.length) {
            pt = Arrays.copyOf(pt, (int) ptLen[0]);
        }
        return pt;
    }

    /** AEAD associated data: utf8(id) followed by big-endian uint64(ts). */
    static byte[] aad(String id, long ts) {
        byte[] idBytes = id.getBytes(StandardCharsets.UTF_8);
        return ByteBuffer.allocate(idBytes.length + 8)
                .order(ByteOrder.BIG_ENDIAN)
                .put(idBytes)
                .putLong(ts)
                .array();
    }

    static byte[] randomNonce() {
        byte[] n = new byte[NONCE_SIZE];
        RANDOM.nextBytes(n);
        return n;
    }

    private static boolean isAllZero(byte[] b) {
        int acc = 0;
        for (byte x : b) {
            acc |= x;
        }
        return acc == 0;
    }

    private static void requireLen(byte[] b, int len, String what) {
        if (b == null || b.length != len) {
            throw new IllegalArgumentException("crypto: " + what + " must be " + len + " bytes");
        }
    }
}

File: sdk/ios/Sources/SecureGatewaySDK/Crypto.swift
import Foundation
import Crypto      // CryptoKit on Apple platforms
import Sodium
import Clibsodium  // direct C API for explicit-nonce XChaCha20-Poly1305 seal

/// E2EE crypto primitives (PRD FR-5), matching the Go reference `internal/e2ee/e2ee.go` and
/// the JVM SDK byte-for-byte. The interop vectors in `internal/e2ee/testdata/vectors.json`
/// are the cross-platform contract.
///
/// - X25519 ECDH via CryptoKit `Curve25519.KeyAgreement` (raw shared secret, low-order
///   points rejected by CryptoKit).
/// - HKDF-SHA256 via CryptoKit `HKDF<SHA256>` (RFC 5869 — matches Go's x/crypto/hkdf).
/// - XChaCha20-Poly1305 (24-byte nonce) via libsodium (swift-sodium); CryptoKit's ChaChaPoly
///   is the 12-byte IETF variant and would NOT match.
public enum Role: String {
    case mobile
    case desktop
}

public struct KeyPair {
    public let privateKey: Data  // 32 bytes
    public let publicKey: Data   // 32 bytes
}

public enum CryptoError: Error, Equatable {
    case ecdhFailed
    case sealFailed
    case openFailed
    case badLength
    case replay // SG-02: envelope id already delivered on this session
    case stale  // SG-02: envelope ts is outside the replay window
}

public enum Crypto {
    public static let keySize = 32
    public static let nonceSize = 24
    static let infoPrefix = "secure-gateway/e2ee/v2|"

    private static let sodium = Sodium()

    public static func generateKeyPair() -> KeyPair {
        let priv = Curve25519.KeyAgreement.PrivateKey()
        return KeyPair(privateKey: priv.rawRepresentation, publicKey: priv.publicKey.rawRepresentation)
    }

    public static func publicFromPrivate(_ priv: Data) throws -> Data {
        let key = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: priv)
        return key.publicKey.rawRepresentation
    }

    /// Raw X25519 shared secret (matches Go's curve25519.X25519).
    public static func sharedSecret(myPriv: Data, peerPub: Data) throws -> Data {
        let priv = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: myPriv)
        let pub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerPub)
        let secret = try priv.sharedSecretFromKeyAgreement(with: pub)
        return secret.withUnsafeBytes { Data($0) }
    }

    /// HKDF-SHA256 over the keying material, info = "secure-gateway/e2ee/v2|"+dir.
    static func deriveKey(ikm: Data, salt: Data, dir: String) -> Data {
        let info = Data((infoPrefix + dir).utf8)
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ikm),
            salt: salt,
            info: info,
            outputByteCount: keySize)
        return key.withUnsafeBytes { Data($0) }
    }

    /// AEAD associated data: utf8(id) || big-endian uint64(ts).
    static func aad(id: String, ts: Int64) -> Data {
        var d = Data(id.utf8)
        var be = UInt64(bitPattern: ts).bigEndian
        withUnsafeBytes(of: &be) { d.append(contentsOf: $0) }
        return d
    }

    /// XChaCha20-Poly1305 seal with an explicit nonce. Returns ciphertext+tag (no nonce).
    /// swift-sodium's Swift `encrypt` always generates its own nonce, so we call the libsodium C
    /// function directly (mirroring swift-sodium's internal call) to pass our deterministic nonce.
    static func aeadSeal(key: Data, nonce: Data, aad: Data, plaintext: Data) throws -> Data {
        let msg = [UInt8](plaintext)
        let ad = [UInt8](aad)
        let npub = [UInt8](nonce)
        let sk = [UInt8](key)
        let abytes = Int(crypto_aead_xchacha20poly1305_ietf_abytes())
        var ct = [UInt8](repeating: 0, count: msg.count + abytes)
        var ctLen: UInt64 = 0
        let rc = crypto_aead_xchacha20poly1305_ietf_encrypt(
            &ct, &ctLen,
            msg, UInt64(msg.count),
            ad, UInt64(ad.count),
            nil, npub, sk)
        guard rc == 0 else { throw CryptoError.sealFailed }
        return Data(ct.prefix(Int(ctLen)))
    }

    static func aeadOpen(key: Data, nonce: Data, aad: Data, cipher: Data) throws -> Data {
        guard let pt = sodium.aead.xchacha20poly1305ietf.decrypt(
            authenticatedCipherText: [UInt8](cipher),
            secretKey: [UInt8](key),
            nonce: [UInt8](nonce),
            additionalData: [UInt8](aad)) else {
            throw CryptoError.openFailed
        }
        return Data(pt)
    }

    static func randomNonce() -> Data {
        var bytes = [UInt8](repeating: 0, count: nonceSize)
        _ = sodium.randomBytes.buf(length: nonceSize).map { bytes = $0 }
        return Data(bytes)
    }
}
